### PR TITLE
Harden DASC recalibration lifecycle

### PR DESCRIPTION
**Pull request by kaix-nv**
Follow-up to #2377 addressing its Claude inline lifecycle review.

Changes:
- accept ModelOpt dynamic subclasses through supported GDN base classes while rejecting unrelated lookalikes;
- make both structural and decay staleness non-fatal for checkpoint save and mode composition while keeping export strict;
- restore repeat-mode protection and make public recalibration replace the existing DASC state entry in place, collapsing any provisional duplicates;
- validate FP16/BF16 casts by re-derived horizons plus the exact selected-head mask; retain the BF16 decay digest only as calibration provenance;
- inherit base Pydantic settings and override only protected_namespaces;
- document and test the lifecycle contract.

Validation:
- 17 focused DASC tests pass
- 128 combined DASC, weight-sparsity, and attention-mode tests pass
- pre-commit passes
- real Qwen3NextGatedDeltaNet policy smoke test passes

Signed-off-by: Kai Xu <kaix@nvidia.com>

## Summary by CodeRabbit

* **New Features**
  * Recalibrating DASC now replaces the previous calibration state instead of creating duplicate history.
  * Added support for additional GatedDeltaNet-based model variants, including dynamically generated variants.
  * Improved compatibility with models saved using lower-precision formats.

* **Bug Fixes**
  * Improved validation of calibration metadata and retained-head selections.
  * Prevented unsupported models that only resemble supported implementations from being accepted.

* **Documentation**
  * Clarified supported model types and recalibration behavior.

### CODE REVIEW

File: modelopt/torch/sparsity/state_sparsity/policy.py [hunk 1]
```diff
@@ -62,9 +63,9 @@ def compute_gdn_decay_horizons(
 
 
 def _is_gdn_module(module: nn.Module) -> bool:
-    """Return whether a module has one of the explicitly supported GDN implementations."""
+    """Accept supported GDN implementations and their ModelOpt dynamic subclasses."""
     return (
-        type(module).__name__ in _SUPPORTED_GDN_CLASS_NAMES
+        any(base.__name__ in _SUPPORTED_GDN_CLASS_NAMES for base in type(module).__mro__)
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py:68]: Addressed in #2380. Class recognition now resolves the installed Megatron and Transformers GDN class objects and compares exact identities. Same-name lookalikes are rejected by a regression test.

File: modelopt/torch/sparsity/state_sparsity/conversion.py [hunk 2]
```diff
@@ -98,8 +98,8 @@ def restore_dasc_model(model: nn.Module, config: DASCConfig, metadata: MetadataD
 def update_dasc_metadata(model: nn.Module, config: DASCConfig, metadata: MetadataDict) -> None:
     """Refresh metadata without making unrelated ModelOpt save or compose paths unusable."""
     policy = get_attached_dasc_policy(model)
-    validate_dasc_model_structure(model, policy)
     try:
+        validate_dasc_model_structure(model, policy)
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/conversion.py:102]: Addressed in #2380. Restore now attaches structurally stale policy metadata with an actionable warning, so a saved stale checkpoint remains recoverable. Export remains strict and rejects deployment until recalibration; the save/restore/export lifecycle is tested.

File: modelopt/torch/sparsity/state_sparsity/policy.py [hunk 3]
```diff
@@ -31,6 +31,7 @@
 __all__ = ["analyze_gdn_decay", "compute_gdn_decay_horizons"]
 
 _SUPPORTED_GDN_CLASS_NAMES = frozenset({"GatedDeltaNet", "Qwen3NextGatedDeltaNet"})
+# Covers ordinary FP16/BF16 storage casts; the exact retained-head mask below is the semantic gate.
 _HORIZON_DTYPE_CAST_RTOL = 0.05
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py:35]: Addressed in #2380. The fixed 5% tolerance is replaced by an analytic interval derived from one rounding step in the current storage dtype. Tests use values that demonstrably change under both FP16 and BF16 casts.

File: modelopt/torch/sparsity/state_sparsity/policy.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py:68]: Addressed in #2380. Ordinary subclasses are no longer accepted. Only exact supported class identities, or ModelOpt DynamicModule subclasses whose MRO contains one of those exact identities, pass. Arbitrary subclass and same-name lookalike tests were added.

File: modelopt/torch/sparsity/state_sparsity/conversion.py [hunk 4]
```diff
@@ -110,6 +110,35 @@ def update_dasc_metadata(model: nn.Module, config: DASCConfig, metadata: Metadat
...
+def replace_dasc_mode(
+    model: nn.Module,
+    config: DASCConfig,
+    measurements: Iterable[DASCCalibrationMeasurement | dict],
+) -> nn.Module:
+    """Replace existing DASC mode state in place with a newly derived policy."""
+    model = unwrap_model(model, force_unwrap=True)
+    policy = build_dasc_policy(model, config, measurements)
+    manager = ModeloptStateManager(model)
+    manager.update_last_state_before_new_mode(model)
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/conversion.py:122]: Addressed in #2380. In-place DASC recalibration skips update_last_state_before_new_mode when DASC is already the trailing mode, so missing or stale attached metadata cannot block replacement. Refresh occurs only when a non-DASC mode trails DASC; both paths are tested.
